core/config/config_helper.hpp

```diff
@@ -0,0 +1,170 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+#define GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+
+
+#include <string>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+#include "core/config/registry_accessor.hpp"
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * LinOpFactoryType enum is to avoid forward declaration, linopfactory header,
+ * two template versions of parse
+ */
+enum class LinOpFactoryType : int { Cg = 0 };
+
+
+/**
+ * It is only an intermediate step after dispatching the class base type. Each
+ * implementation needs to deal with the template selection.
+ */
+template <LinOpFactoryType flag>
+deferred_factory_parameter<gko::LinOpFactory> parse(
+    const pnode& config, const registry& context,
+    const type_descriptor& td = make_type_descriptor<>());
+
+
+/**
+ * get_stored_obj searches the object pointer stored in the registry by string
+ */
+template <typename T>
+inline std::shared_ptr<T> get_stored_obj(const pnode& config,
+                                         const registry& context)
+{
+    std::shared_ptr<T> ptr;
+    using T_non_const = std::remove_const_t<T>;
+    ptr = detail::registry_accessor::get_data<T_non_const>(context,
+                                                           config.get_string());
+    GKO_THROW_IF_INVALID(ptr.get() != nullptr, "Do not get the stored data");
+    return ptr;
+}
+
+
+/**
+ * Build the factory from config (map) or search the pointers in
+ * the registry by string.
+ */
+template <typename T>
+deferred_factory_parameter<T> build_or_get_factory(const pnode& config,
```

Comment:
how fixed is the name? I would prefer 
```suggestion
deferred_factory_parameter<T> parse_or_get_factory(const pnode& config,
```
since it calls `parse` not `build` internally.